File: mobile/src/main/java/packi/day/notification/Alarm.java
package packi.day.notification;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.PowerManager;
import android.preference.PreferenceManager;

import java.util.Calendar;
import java.util.Locale;

public class Alarm extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        if (pm == null) return;
        PowerManager.WakeLock wl = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "");

        SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context);
        Boolean isActive = sharedPrefs.getBoolean("notification", false);

        if (isActive) {
            try {
                wl.acquire(1000);
                context.startService(new Intent(context, NotificationExecutor.class));
            } finally {
                wl.release();
            }
        }
    }

    public void init(Context context) {
        AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        if (am == null) return;

        SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context);

        String savedDate = sharedPrefs.getString("dateNotification", "00:00");
        Calendar calendar = Calendar.getInstance(Locale.getDefault());
//        int hours = TimePickerPreferences.getHour(savedDate);
//        int minutes = TimePickerPreferences.getMinute(savedDate);
        int year = calendar.get(Calendar.YEAR);
        int month = calendar.get(Calendar.MONTH);
        int day = calendar.get(Calendar.DATE);

        calendar.set(year, month, day, 0, 0, 0);
        if (calendar.before(Calendar.getInstance(Locale.getDefault()))) {
            calendar.add(Calendar.DATE, 1);
        }

        Intent i = new Intent(context, Alarm.class);
        PendingIntent pi = PendingIntent.getBroadcast(context, 0, i, 0);
        am.setRepeating(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), 1000 * 60 * 60 * 24, pi);
    }

    public void cancelAlarm(Context context) {
        AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        if (am == null) return;

        Intent intent = new Intent(context, Alarm.class);
        PendingIntent sender = PendingIntent.getBroadcast(context, 0, intent, 0);

        am.cancel(sender);
    }
}

File: mobile/src/main/java/packi/day/notification/Alarm.kt
package packi.day.notification

import android.app.AlarmManager
import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.os.PowerManager
import packi.day.store.UserSetting
import java.util.*

class Alarm : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (!UserSetting(context).shouldNotify) return

        val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        val wl = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "")
        try {
            wl.acquire(1000)
            context.startService(Intent(context, NotificationExecutor::class.java))
        } finally {
            wl.release()
        }
    }

    fun init(context: Context) {
        val calendar = Calendar.getInstance(Locale.getDefault())

        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH)
        val day = calendar.get(Calendar.DATE)

        val settings = UserSetting(context)

        calendar.set(year, month, day, settings.notificationDate.first.hours, settings.notificationDate.second.minutes, 0)
        if (calendar.before(Calendar.getInstance(Locale.getDefault()))) {
            calendar.add(Calendar.DATE, 1)
        }

        val intent = Intent(context, Alarm::class.java)
        val pendingIntent = PendingIntent.getBroadcast(context, 0, intent, 0)

        val alarm = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarm.setRepeating(AlarmManager.RTC_WAKEUP, calendar.timeInMillis, (1000 * 60 * 60 * 24).toLong(), pendingIntent)
    }

    fun cancelAlarm(context: Context) {
        val intent = Intent(context, Alarm::class.java)
        val sender = PendingIntent.getBroadcast(context, 0, intent, 0)

        val am = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        am.cancel(sender)
    }
}


class AlarmService : Service() {

    private val alarm = Alarm()

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        alarm.init(this)
        return Service.START_STICKY
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }
}

class AutoStart : BroadcastReceiver() {

    private val alarm = Alarm()

    override fun onReceive(context: Context, intent: Intent) {
        if ("android.intent.action.BOOT_COMPLETED" != intent.action) return

        if (UserSetting(context).shouldNotify) {
            alarm.init(context)
        } else {
            alarm.cancelAlarm(context)
        }

    }
}